main/core.js

```diff
@@ -122,34 +126,46 @@ async function start (ctx) {
   /** @type {string | null} */
   let exitReason = null
 
-  app.on('before-quit', () => {
-    childProcess.kill()
-  })
-
-  childProcess.on('close', code => {
-    console.log(`Core closed all stdio with code ${code ?? '<no code>'}`)
-
-    if (code === 2) {
-      // FIL_WALLET_ADDRESS did not pass our screening. There is not much
-      // we can do about that, there is no point in reporting this error
-      // to Sentry.
-      return
-    }
+  const controller = new AbortController()
+  const { signal } = controller
 
-    Sentry.captureException('Core exited', scope => {
-      // Sentry UI can't show the full 100 lines
-      scope.setExtra('logs', logs.getLastLines(10))
-      scope.setExtra('reason', exitReason)
-      return scope
-    })
+  const onBeforeQuit = () => childProcess.kill()
+  app.on('before-quit', onBeforeQuit)
+  signal.addEventListener('abort', () => {
+    app.removeListener('before-quit', onBeforeQuit)
   })
 
-  childProcess.on('exit', (code, signal) => {
-    const reason = signal ? `via signal ${signal}` : `with code: ${code}`
-    const msg = `Core exited ${reason}`
-    console.log(msg)
-    exitReason = signal || code ? reason : null
-  })
+  await Promise.all([
+    (async () => {
+      const [code, exitSignal] = await once(childProcess, 'exit', { signal })
+      controller.abort()
+      const reason = exitSignal
+        ? `via signal ${exitSignal}`
+        : `with code: ${code}`
+      const msg = `Core exited ${reason}`
+      console.log(msg)
+      exitReason = exitSignal || code ? reason : null
+    })(),
+    (async () => {
+      const [code] = await once(childProcess, 'close')
+      controller.abort()
+      console.log(`Core closed all stdio with code ${code ?? '<no code>'}`)
```

Comment:
If I remember correctly my assumptions behind the original implementation: We rely on the fact that `exit` is emitted first and `close` is emitted second. That allows us to capture `exitReason` in the `exit` event handler and later use it in the `close` handler.

If that assumption is correct, then we don't need `Promise.all` and can simplify the code to listen for these events sequentially.

```js
const [code, exitSignal] = await once(childProcess, 'exit', { signal })
// ...
const [closeCode] = await once(childProcess, 'close')
// ...
```

WDYT?

Reply:
I tried it like this, then force-quit the core process, and only this got logged:
 
```
Core exited via signal SIGKILL
```

And if I move the `close` handler first, only that gets logged.

I can't tell you exactly why, but unfortunately this approach doesn't work.

Reply:
Oh, maybe we added the other event listener too late after the event was already emitted.

```js
const onceExited = once(childProcess, 'exit', { signal })
const onceClosed = once(childProcess, 'close')

const [code, exitSignal] = await onceExited
// ...
const [closeCode] = await onceClosed
// ...
```

At this point, I am not sure if this is an improvement over your `Promise.all` solution. WDYT?

Reply:
No that works, I think I prefer it :)

Reply:
https://github.com/filecoin-station/desktop/pull/1549/commits/aeb6f82f637b9b0d89533ae3257e7646a92c663e